R/db-interface.R

```diff
@@ -201,20 +195,36 @@ copy_dm_to <- function(dest, dm, ...,
     return(dm)
   }
 
-  copy_data <- build_copy_data(dm, dest, table_names_out, set_key_constraints, dest_con)
+  queries <- build_copy_queries(dest_con, dm, set_key_constraints, temporary, table_names_out)
 
-  ticker <- new_ticker("uploading data", nrow(copy_data), progress = progress)
-  new_tables <- pmap(
-    copy_data, ticker(copy_to),
-    dest = dest,
-    temporary = temporary,
-    overwrite = FALSE,
-    ...
+  # create tables
+  walk(queries$create_table_queries$sql, ~{
+    rs <- DBI::dbSendStatement(dest_con, .x)
+    DBI::dbClearResult(rs)
+  })
+
+  # populate tables
+  pwalk(
+    queries$create_table_queries[c("table", "remote_table")],
+    ~  {
+      DBI::dbAppendTable(dest_con, DBI::SQL(.y), dm[[.x]])
```

Comment:
To fix the failures on SQL Server, we need to ensure that `dbExistsTable()` works for the table names we create. The implementation in {odbc} uses `dbUnquoteIdentifier()`: https://github.com/r-dbi/odbc/blob/main/R/Table.R#L204-L224 . As a workaround we could try `dbWriteTable(append = TRUE)`:

```suggestion
      # https://github.com/r-dbi/odbc/issues/480
      DBI::dbWriteTable(dest_con, DBI::SQL(.y), dm[[.x]], append = TRUE)
```

Internally, `odbc::dbAppendTable()` calls `dbWriteTable()` anyway...